DRA: CRUD conformance tests

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Promoting real tests (https://github.com/kubernetes/kubernetes/pull/133132) turned out to be harder than expected (should be rewritten to be self-contained, additional reviews, etc.).

They would not achieve 100% endpoint+operation coverage because real tests only use some of the operations. Therefore each API type has to be covered with CRUD-style tests which only exercise the apiserver, then maybe additional functional tests can be added later (depending on time and motivation).

#### Which issue(s) this PR is related to:

https://github.com/kubernetes/kubernetes/issues/114183
https://github.com/kubernetes/kubernetes/issues/133691
KEP: https://github.com/kubernetes/enhancements/issues/4381

#### Special notes for your reviewer:

The machinery for testing different API types is meant to be reusable, so it gets added in the new e2e/framework/conformance helper package.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
DRA: CRUD conformance tests

Promoting real tests turned out to be harder than expected (should be rewritten
to be self-contained, additional reviews, etc.).

They would not achieve 100% endpoint+operation coverage because real tests only
use some of the operations. Therefore each API type has to be covered with
CRUD-style tests which only exercise the apiserver, then maybe additional
functional tests can be added later (depending on time and motivation).

The machinery for testing different API types is meant to be reusable, so it
gets added in the new e2e/framework/conformance helper package.

pohly commented:
/test pull-kubernetes-e2e-kind

dims commented:
> But as these tests are new, I reverted back to normal f.It. The promotion can follow later after the usual 2 week period.

thanks for changing it back to `f.It`. Yes please, let's watch it for 2 weeks and do a follow up PR to promotion.

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,442 @@
+	// Delete the resource.
+	ginkgo.By(fmt.Sprintf("Deleting %s", klog.KObj(initialObject)))
+	err = client.Delete(ctx, initialObject.GetName(), metav1.DeleteOptions{})
+	framework.ExpectNoError(err, "delete updated resource")
```
we don't control end-user environments ... *we* don't add finalizers, but something in a user-cluster could, and if their finalizer is well-behaved, shouldn't make them fail conformance (we've seen this happen with other resources, e.g. https://github.com/kubernetes/kubernetes/pull/96742)

I think tolerating a brief finalizer → delete delay is a good idea.

pohly commented on test/e2e/framework/conformance/conformance.go:
This helper is meant to be the go-to solution for CRUD tests of new types. Let's make sure that it gets reviewed thoroughly before merging.

Example output for a namespaced resource with status:
```
[sig-node] [DRA] CRUD Tests resource.k8s.io/v1 ResourceClaim [sig-node, DRA]
/nvme/gopath/src/k8s.io/kubernetes/test/e2e/dra/dra.go:119
  STEP: Creating a kubernetes client @ 08/29/25 08:30:31.651
  I0829 08:30:31.651203 63964 util.go:454] >>> kubeConfig: /var/run/kubernetes/admin.kubeconfig
  STEP: Building a namespace api object, basename dra @ 08/29/25 08:30:31.651
  STEP: Waiting for a default service account to be provisioned in namespace @ 08/29/25 08:30:31.66
  STEP: Waiting for kube-root-ca.crt to be provisioned in namespace @ 08/29/25 08:30:31.664
  STEP: Get resource.k8s.io/v1 @ 08/29/25 08:30:31.669
  STEP: Creating:
     apiVersion: resource.k8s.io/v1
     kind: ResourceClaim
     metadata:
       name: example
       namespace: dra-6943
     spec:
       devices:
         requests:
         - exactly:
             deviceClassName: dra.example.com
           name: req-0
      @ 08/29/25 08:30:31.671
  STEP: Getting dra-6943/example @ 08/29/25 08:30:31.675
  STEP: Updating:
     apiVersion: resource.k8s.io/v1
     kind: ResourceClaim
     metadata:
       labels:
         test.dra.example.com: test
       name: example
       namespace: dra-6943
       resourceVersion: "467"
       uid: f9bab3c8-62b6-4c26-97ee-4231eb34a9d5
     spec:
       devices:
         requests:
         - exactly:
             deviceClassName: dra.example.com
           name: req-0
     status:
       allocation: {}
      @ 08/29/25 08:30:31.678
  STEP: Updating status:
     apiVersion: resource.k8s.io/v1
     kind: ResourceClaim
     metadata:
       labels:
         test.dra.example.com: test
       name: example
       namespace: dra-6943
       resourceVersion: "468"
       uid: f9bab3c8-62b6-4c26-97ee-4231eb34a9d5
     spec:
       devices:
         requests:
         - exactly:
             deviceClassName: dra.example.com
           name: req-0
     status:
       allocation: {}
      @ 08/29/25 08:30:31.683
  STEP: Getting dra-6943/example status @ 08/29/25 08:30:31.686
  STEP: Deleting dra-6943/example @ 08/29/25 08:30:31.69
  STEP: Creating again:
     apiVersion: resource.k8s.io/v1
     kind: ResourceClaim
     metadata:
       name: example
       namespace: dra-6943
     spec:
       devices:
         requests:
         - exactly:
             deviceClassName: dra.example.com
           name: req-0
      @ 08/29/25 08:30:31.696
  STEP: Patching:
     apiVersion: resource.k8s.io/v1
     kind: ResourceClaim
     metadata:
       labels:
         test.dra.example.com: test
       name: example
       namespace: dra-6943
     status:
       allocation: {}
      @ 08/29/25 08:30:31.7
  STEP: Patching status:
     apiVersion: resource.k8s.io/v1
     kind: ResourceClaim
     metadata:
       labels:
         test.dra.example.com: test
       name: example
       namespace: dra-6943
     status:
       allocation: {}
      @ 08/29/25 08:30:31.705
  STEP: Listing resource.k8s.io/v1, Resource=resourceclaims collection with label selector e2e-test.kubernetes.io=dra-6943 @ 08/29/25 08:30:31.713
  STEP: Listing resource.k8s.io/v1, Resource=resourceclaims without namespace and with label selector e2e-test.kubernetes.io=dra-6943 @ 08/29/25 08:30:31.717
  STEP: Deleting resource.k8s.io/v1, Resource=resourceclaims collection with label selector e2e-test.kubernetes.io=dra-6943 @ 08/29/25 08:30:31.72
  I0829 08:30:31.732857 63964 helper.go:125] Waiting up to 7m0s for all (but 0) nodes to be ready
  STEP: Destroying namespace "dra-6943" for this suite. @ 08/29/25 08:30:31.736
```

pohly committed:
DRA E2E: only check ResourceClaims in the test namespace

That was the original intent, but the implementation then ended up checking
ResourceClaims in all namespaces. Depending on timing this was merely
misleading (showing ResourceClaim changes from a different test running in
parallel), but with upcoming CRUD tests which intentionally set an allocation
result without a finalizer it breaks the non-CRUD tests when they check the
those CRUD ResourceClaims.

dims commented:
/assign @BenTheElder @liggitt

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,464 @@
+	config := dynamic.ConfigFor(f.ClientConfig())
+	httpClient, err := rest.HTTPClientFor(config)
+	framework.ExpectNoError(err, "construct HTTP client")
+	restClient, err := rest.UnversionedRESTClientForConfigAndClient(config, httpClient)
```
do we have to use a raw REST client directly? can we use a discovery client to fetch discovery endpoints for this?

liggitt replied:
this is fine, I see us directly requesting discovery endpoints in other places

pohly commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,464 @@
+	// name and namespace updated by the test.
+	// The update object is created by adding all fields from
+	// the patch object to the initial object.
+	var initialObject *unstructured.Unstructured
+	err = yaml.Unmarshal([]byte(tc.InitialData), &initialObject)
+	framework.ExpectNoError(err, "decode initialData")
+	if tc.Namespaced {
+		initialObject.SetNamespace(f.Namespace.Name)
+	} else {
+		initialObject.SetName(initialObject.GetName() + "-" + f.Namespace.Name)
+	}
```
Done.

pohly commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,464 @@
+	ginkgo.By(fmt.Sprintf("Get %s", gv))
+	path := "/apis/" + gv.String()
+	var api unstructured.Unstructured
+	err = restClient.
+		Get().
+		AbsPath(path).
```
Do you want to avoid setting the path based on the gvr? Why? Because the test should verify that discovery works or because the path might not follow the pattern expected here?

It could be changed, but not done yet.

liggitt replied:
mmm, I guess directly requesting the exact path for discovery and decoding the response into the expected discovery structs is less magical than using an encapsulating discovery client

pohly commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,464 @@
+	framework.ExpectNoError(err, "delete updated resource")
+	ensureNotFound(ctx, getResource)
+
+	// Recreate for patching.
```
Ack.

pohly commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,464 @@
+		verifyObject("patch status", updateObject, existingObject)
+	}
+
+	// Add a label, then use that label as selector in List and DeleteCollection calls.
```
It's now expected to be included in the initial object, with label name/value chosen by the common helper.

liggitt reviewed (commented):
Did another sweep. I like how succinct the use is. Let's see if we can protect against typos / errors / omissions leading to unexpected default behavior or silently dropping stuff.

pohly commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,724 @@
+	GetUpdateSpec(object *unstructured.Unstructured) *unstructured.Unstructured
+
+	// GetUpdateStatus modifies the status of the initial object with
+	// the spec updated.
```
What I meant to explain was what the method gets called for (initial object + spec update).

Changed to this:
```
	// GetUpdateSpec modifies an existing object.
	// It gets called for the result of creating the initial object.
...

	// GetUpdateStatus modifies the status of an existing object.
	// It gets called for the result of creating the initial object
	// and then updating its spec.
```

pohly commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,564 @@
+	//
+	// The name of cluster-scoped resources gets extended with
+	// `-<test namespace name>` to make it unique.
+	Namespaced bool
```
I made it a pointer, with an early failure at runtime if unset ("Test case error: Namespaced must be set").

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,724 @@
+
+	// StrategicMergePatchSpec must contain modified fields for InitialSpec
+	// for use in a strategic merge patch, using JSON encoding.
+	// The result of applying the patch must be the same as in UpdateSpec.
```
sounds good, thanks

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,845 @@
+// The data is used like this:
+// - create InitialSpec -> update with UpdateSpec -> update status with UpdateStatus
+// - create InitialSpec -> apply StrategicMergePatchSpec and compare against UpdateSpec -> apply StrategicMergePatchStatus and compare against UpdateStatus
+type ResourceTestcase[T any] struct {
```
not sure if it makes it easier or harder to work with, but I _think_ you could constrain T to be an implementation of runtime.Object, and then deal with `T` throughout instead of `*T`

```suggestion
type ResourceTestcase[T runtime.Object] struct {
```

(non-blocking, feel free to do it if it's easy and works well)

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,845 @@
+// Gomega could be configured to log unstructured.Unstructured as YAML,
+// but right now it isn't, so we have to do that ourselves instead of
+// using format.Object.
+func marshalYAML(obj any, indent int) string {
```
nit: clarify in the method name this is only for logging, not for making yaml to then read / consume

```suggestion
func marshalYAMLForLogging(obj any, indent int) string {
```

pohly replied:
Done.

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,845 @@
+func ensureNotFound(ctx context.Context, get func(context.Context) (*unstructured.Unstructured, error)) {
+	ginkgo.GinkgoHelper()
+	ginkgo.By("Checking for existence")
+	gomega.Eventually(ctx, func(ctx context.Context) error {
```
nit: add a comment explaining this allows objects to be finalized asynchronously before actually being removed

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,845 @@
+	}
+
+	// Patch the resource, potentially using multiple different patch types.
+	// The result must be the same each time.
```
```suggestion
	// The result must be the same each time if VerifyContent() is true.
```

liggitt reviewed (commented):
lgtm, just nits and godoc suggestions at this point

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,845 @@
+
+	// Same for the status. In addition, read the status (same result, but different endpoint+method).
+	if tc.HasStatus() {
+		mustGet := false
+		gomega.Eventually(ctx, func(ctx context.Context) error {
+			if mustGet {
+				ginkgo.By(fmt.Sprintf("Getting updated %s", klog.KObj(desiredInitialObject)))
+				existingObject, err = client.Get(ctx, desiredInitialObject.GetName(), metav1.GetOptions{})
```
I think we always want to exercise a get of the status subresource to make sure that works

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,845 @@
+	if index < 0 {
+		framework.Failf("API for %s does not include entry for %s, got:\n%s", gv, resource, marshalYAML(&api, 1))
+	}
+
```
I just realized I think we're missing exercising watch.

I'd recommend:
1. starting label-selector-filtered dynamic informer(s) that handle rewatching if needed (a global one, and a namespaced one for namespaced resources) at the start of the test, wait for them to sync, and accumulate the events each observe
2. after the create, assert watchers observed creates
3. after the update, assert watchers observed updates
4. after the delete, assert watchers observed deletes
5. terminate the informers/watchers in test cleanup

pohly replied:
@liggitt: you said on Slack that you are sure that apisnoop would complain about not covering watches.

But "watch" isn't listed in https://github.com/kubernetes/kubernetes/blob/0969bb173cbbd74ea1e87cd8646746f414fc58ca/test/conformance/testdata/pending_eligible_endpoints.yaml#L2-L30 and apisnoop doesn't seem to care either:

https://apisnoop.cncf.io/conformance-progress/endpoints/1.35.0/?filter=untested

liggitt replied:
> I'm not sure about checking for exact events. Can we be sure that we get exactly those?

Yes, the events I described *must* be observed, even in the presence of watch restarts.

> after the create, assert watchers observed 1 added event (followed by 0..* update events)

A newly started informer will see an initial Added event for all objects, and we can rely on that and should verify that. If other writes are being done to the object, there can be 0..* Update events following the Add, and we should tolerate seeing those.

> after the update, assert watchers observed 1..* update events

There must be at least one update event observed in response to our update, even if the watch was restarted or a relist was required. Other updates from other actors could occur, and we should tolerate seeing those.

> after the delete, assert watchers observed 0..* update events followed by 1 delete event

There _must_ be a delete event observed after our delete (it might be a DeletionFinalStateUnknown object, which we should tolerate). If our watch is filtered to the specific object name and label selector, the only way this would not be the case is if someone outside the test recreated the exact named object with the exact test-specific label after we deleted the object, but that would also break the "check that the object is not found" poll after deletion.

It is possible there are preceding update events (if there were concurrent updaters, or if the object used graceful deletion, or if the object had a finalizer that needed to be removed first), so we should tolerate seeing those before the delete.

liggitt replied:
> IMHO it's better to check for the outcome: eventually an informer cache has the desired object. I implemented that.

I think we should check for the specific types of events observed, at the points I indicated (after create, after update, after delete)

liggitt replied:
Yeah, I agree we need the robustness of an informer that will rewatch/relist if needed.

pohly replied:
Updated.

`conformance.go` started to get too big, so I split it up into several different files, each with a different focus.

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,986 @@
+		} else {
+			gomega.Expect(actual.GetName()).Should(gomega.Equal(expected.GetName()), "%s: name in returned object", what)
+			gomega.Expect(actual.GetNamespace()).Should(gomega.Equal(expected.GetNamespace()), "%s: namespace in returned object", what)
+		}
```
Add a check that the resourceVersion invariant from KEP-5504 is valid

```
gomega.Expect(actual).To(apimachineryutils.HaveValidResourceVersion())
```

This will ensure all future conformance-tested APIs continue to have resourceVersions that match documented behavior for Kubernetes APIs

pohly replied:
Saw this too late, not included in latest update yet. Will do next...

pohly replied:
Added.

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,986 @@
+		framework.ExpectNoError(err, "delete %s", resourceType)
+		ensureNotFound(ctx, getResource)
+	})
+	verify("create", desiredInitialObject, existingObject)
```
yes, I'm *really* looking forward to never reviewing the details of another CRUD conformance test again :)

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,896 @@
+	})
+	verify("create", desiredInitialObject, existingObject,
+		// Initial creation of the object followed by some optional updates by cluster components.
+		matchEvents("^add %[1]s, %[2]s\n(update %[1]s, %[2]s\n)*$", existingObject),
```
Routing the event aggregation through stringification and regex matching is pretty hard to follow. I also don't think the string substitution of object name and uid into the regex is handling escaping at all.

Since we're labeling the objects uniquely, and our watchers will only see events for a single object, it seems like we can make `matchEvents` directly check that all objects in all events match existingObject name/namespace/uid, and just make the custom regex bit verify the event types:

```
// add followed by any number of updates
matchEvents(existingObject, "^add(,update)*$")
...
// at least one update, followed by any number of additional updates
matchEvents(existingObject, "^update(,update)*$")
...
// any number of optional updates followed by a delete
matchEvents(existingObject, "^(update,)*delete$")
```

Then we don't need multi-line matching or string substitution into the regexes.

BenTheElder commented:
>   [FAIL] [sig-apps] ControllerRevision [Serial] [It] should manage the lifecycle of a ControllerRevision [Conformance] [sig-apps, Serial, Conformance]

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133747/pull-kubernetes-conformance-image-test/1973394346661122048

oof .... do not like flaking conformance tests (unrelated to this PR)

it's possible this job config needs fixup

BenTheElder commented:
/retest

michaelasp commented:
/retest
The fix was merged after the trigger

pohly committed:
test: format unstructured as YAML

It's a nested map which looks a lot nicer as YAML, in particular
when it represents a Kubernetes object.

Unit+integration tests using ktesting+gomega and E2E tests benefit from this
change.

pohly commented:
I had to rebase to get access to `apimachineryutils.HaveValidResourceVersion`. https://github.com/kubernetes/kubernetes/compare/839045e442ff485ce395d4fb90901d83e8794fc6..2500c4d0c2811b6be8ac9a2ea5f2f43e5b1e33b2 is that rebase with no content changes.

The actual changes are in https://github.com/kubernetes/kubernetes/compare/2500c4d0c2811b6be8ac9a2ea5f2f43e5b1e33b2..8093062d4cc36bd8dc3be4c6e44ef13a2b667ec3

- checking ResourceVersion
- simplifying the event sequence check
- `unstructured.Unstructured` rendering as YAML in Gomega output

Small cleanups in https://github.com/kubernetes/kubernetes/compare/8093062d4cc36bd8dc3be4c6e44ef13a2b667ec3..d9df6f82a77e659855720ea6f4be01967e2874f3

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,940 @@
+			framework.ExpectNoError(err, "access meta data")
+			gomega.Expect(metaData).To(apimachineryutils.HaveValidResourceVersion())
+
+			nextRV, ok := new(big.Int).SetString(metaData.GetResourceVersion(), 10)
+			if !ok {
+				framework.Failf("Parse ResourceVersion %s as int", metaData.GetResourceVersion())
+			}
+			if nextRV.Cmp(rv) < 0 {
```
don't parse and compare with big.Int here, keep them as strings, use the apimachinery method made for this purpose, and assert >= 0 and no error

https://github.com/kubernetes/kubernetes/blob/8ac5701d3a146b367c83449d1eb6d872c003752e/staging/src/k8s.io/apimachinery/pkg/util/resourceversion/resourceversion.go#L32-L45

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,940 @@
+	verify("create", desiredInitialObject, existingObject,
+		// Initial creation of the object followed by some optional updates by cluster components.
+		matchEvents(existingObject, "^add,(update,)*$"),
+	)
```
save `createdResourceVersion := existingObject.GetResourceVersion()` here, and compare specifically with the resource version after `verify("update"...` to make sure the resource version *increased*

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,940 @@
+		for _, e := range events.Events {
+			checkNext(e.oldObj)
+			checkNext(e.newObj)
+
```
In addition to checking the rvs in isolation to make sure they never *decrease*, we want to verify *somewhere* that the RV on the updated object is > the RV on the created object

pohly replied:
Added as suggested in https://github.com/kubernetes/kubernetes/pull/133747#discussion_r2399032778.

pohly commented:
/test pull-kubernetes-unit

https://github.com/kubernetes/kubernetes/issues/134386

/test pull-kubernetes-node-e2e-containerd-2-0-dra-alpha-beta-features pull-kubernetes-e2e-gce

Cluster resp. machine didn't come up. I know that Ben wants me to investigate this because someone has to, but I don't have the time right now to learn about that stuff... :grimacing:

liggitt commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,940 @@
+			checkNext(e.oldObj)
+			checkNext(e.newObj)
+
+			// Reset after delete?
```
it is when watching a single object:
* relist/resyncs will just replay the latest observed thing
* delete/recreate will result in later RVs

liggitt commented:
I think there are probably opportunities to simplify the framework a little to make it easier to follow, but I'm happy with the current state of what this is testing. It represents the best practices we know of from other conformance tests (tolerate non-instantanous deletes with finalizers, use robust informer-driven checks of watch event invariants, etc).

/lgtm
/approve

BenTheElder commented on test/e2e/dra/dra.go:
```diff
@@ -77,6 +78,156 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), func() {
+						},
+					},
+					UpdateSpec: func(obj *resourceapi.DeviceClass) *resourceapi.DeviceClass {
+						obj.Spec.Selectors[0].CEL.Expression = "1 == 0" // Still matches no devices.
```
heh. cc @jpbetz @cici37

BenTheElder commented on test/e2e/framework/conformance/conformance.go:
```diff
@@ -0,0 +1,946 @@
+	//
+	// This should not be enabled in conformance tests because admission
+	// is allowed to modify what is being stored.
+	ContentVerificationEnabled bool
```
Seems like there's room to validate this struct, e.g. checking that if this bool is set then StrategicMergePatchSpec needs to be set if UpdateSpec is set?

I like what we're doing here overall, but it seems like if you don't copy-paste from another test, it might be easy to use incorrectly, even though it's still easier than writing these from scratch.

BenTheElder reviewed (commented):
Agree to merge and iterate from here, I think there's room to make this even slightly more guarded against misuse and simplify a bit but overall I think this will be a big improvement over hand-writing this style of test

/lgtm
/approve

liggitt commented:
```
{Failed  === RUN   TestNamespaceScopedWatch/namespaced_watch,_request_with_name,_request_without_namespace,_field_selector_with_metadata.name_and_spec.nodeName
    utils.go:158: incorrect event:   watch.Event{
        - 	Type: "ADDED",
        + 	Type: "ERROR",
        - 	Object: &example.Pod{
        - 		ObjectMeta: v1.ObjectMeta{Name: "t12-foo1", Namespace: "t12-ns1", ResourceVersion: "50"},
        - 		Spec:       example.PodSpec{NodeName: "t12-bar1"},
        - 	},
        + 	Object: s`&Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,RemainingItemCount:nil,},Status:Failure,Message:Internal error occurred: etcd event received with PrevKv=nil (key="/pods/t12-ns1/t12-foo1", modRevision=49, type=PUT),Reason:InternalError,Detail`...,
          }
```
fyi @serathius 

/retest

aojea commented:
The integration failure is 

https://github.com/kubernetes/kubernetes/issues/133861

The e2e is :
> Kubernetes e2e suite: [It] [sig-node] [DRA] control plane [ConformanceCandidate] supports reusing resources expand_less

liggitt commented:
> The e2e is :
> 
> Kubernetes e2e suite: [It] [sig-node] [DRA] control plane [ConformanceCandidate] supports reusing resources expand_less

hmmm, is the failure any conflict with the DRA resources created by this test?

pohly commented:
> hmmm, is the failure any conflict with the DRA resources created by this test?

Doesn't seem likely. Each DRA test uses a unique driver name and the objects created by these new tests shouldn't interfere with any concurrently running test. Let's keep an eye open for it though before promoting the new tests.

Speaking of which, that promotion is in https://github.com/kubernetes/kubernetes/pull/134439, pending for now until the tests have enough soak time.
